share/eds/p2p/server.go

```diff
@@ -0,0 +1,163 @@
+// This client is run by bridge nodes and full nodes. For more information, see ADR #13
+type Server struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	host       host.Host
+	protocolID protocol.ID
+
+	store *eds.Store
+}
+
+// NewServer creates a new ShrEx/EDS server.
+func NewServer(host host.Host, store *eds.Store) *Server {
+	return &Server{
+		host:       host,
+		protocolID: protocolID,
+		store:      store,
+	}
+}
+
+func (s *Server) Start(context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.host.SetStreamHandler(s.protocolID, s.handleStream)
+	return nil
+}
+
+func (s *Server) Stop(context.Context) error {
+	defer s.cancel()
+	s.host.RemoveStreamHandler(s.protocolID)
+	return nil
+}
+
+func (s *Server) handleStream(stream network.Stream) {
+	log.Debug("server: handling eds request")
+
+	// read request from stream to get the dataHash for store lookup
+	req, err := s.readRequest(stream)
+	if err != nil {
+		log.Errorw("server: reading request from stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	var edsReader io.ReadCloser
+	status := p2p_pb.Status_OK
+	err = share.DataHash(req.Hash).Validate()
+	if err != nil {
+		status = p2p_pb.Status_INVALID
+	} else {
+		// determine whether the EDS is available in our store
+		edsReader, err = s.store.GetCAR(s.ctx, req.Hash)
+		if err != nil {
+			status = p2p_pb.Status_NOT_FOUND
+		} else {
+			defer edsReader.Close()
+		}
+	}
+
+	// inform the client of our status
+	err = s.writeStatus(status, stream)
+	if err != nil {
+		log.Errorw("server: writing status to stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+	// if we cannot serve the EDS, we are already done
+	if status != p2p_pb.Status_OK {
+		stream.Close()
+		return
+	}
+
+	// start streaming the ODS to the client
+	err = s.writeODS(edsReader, stream)
+	if err != nil {
+		log.Errorw("server: writing ods to stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	err = stream.Close()
+	if err != nil {
+		log.Errorw("server: closing stream", "err", err)
+		return
+	}
+}
+
+func (s *Server) readRequest(stream network.Stream) (*p2p_pb.EDSRequest, error) {
+	err := stream.SetReadDeadline(time.Now().Add(readDeadline))
+	if err != nil {
+		log.Warn(err)
+	}
+	req := new(p2p_pb.EDSRequest)
+	_, err = serde.Read(stream, req)
+	if err != nil {
+		return nil, err
+	}
+	if err = stream.CloseRead(); err != nil {
+		log.Error(err)
+	}
+
+	return req, nil
+}
+
+func (s *Server) writeStatus(status p2p_pb.Status, stream network.Stream) error {
+	err := stream.SetWriteDeadline(time.Now().Add(writeDeadline))
+	if err != nil {
+		log.Warn(err)
+	}
+	resp := &p2p_pb.EDSResponse{Status: status}
+	_, err = serde.Write(stream, resp)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Server) writeODS(edsReader io.ReadCloser, stream network.Stream) error {
```

Comment:
We should stream the data instead of sending it all as one message. Streaming implies sending the data in portions..The simplest way of doing so is to use `io.Copy`, which would read data out of ODSReader and write it into the stream.

Reply:
done. thanks for pointing me to io.Copy, that's really cool